agent/structs/config_entry_discoverychain.go

```diff
@@ -943,6 +997,76 @@ type ServiceResolverFailover struct {
 	Datacenters []string `json:",omitempty"`
 }
 
+// LoadBalancer determines the load balancing policy and configuration for services
+// issuing requests to this upstream service.
+type LoadBalancer struct {
+	// Policy is the load balancing policy used to select a host
+	Policy string `json:",omitempty"`
+
+	// RingHashConfig contains configuration for the "ring_hash" policy type
+	RingHashConfig RingHashConfig `json:",omitempty" alias:"ring_hash_config"`
+
+	// LeastRequestConfig contains configuration for the "least_request" policy type
+	LeastRequestConfig LeastRequestConfig `json:",omitempty" alias:"least_request_config"`
+
+	// HashPolicies is a list of hash policies to use for hashing load balancing algorithms.
+	// Hash policies are evaluated individually and combined such that identical lists
+	// result in the same hash.
+	// If no hash policies are present, or none are successfully evaluated,
+	// then a random backend host will be selected.
+	HashPolicies []HashPolicy `json:",omitempty" alias:"hash_policies"`
+}
+
+func (l LoadBalancer) IsHashBased() bool {
+	switch l.Policy {
+	case "maglev", "ring_hash":
+		return true
+	default:
+		return false
+	}
+}
+
+// RingHashConfig contains configuration for the "ring_hash" policy type
+type RingHashConfig struct {
+	// MinimumRingSize determines the minimum number of hashes per destination host
+	MinimumRingSize uint64 `json:",omitempty" alias:"minimum_ring_size"`
+
+	// MaximumRingSize determines the maximum number of hashes per destination host
+	MaximumRingSize uint64 `json:",omitempty" alias:"maximum_ring_size"`
+}
+
+// LeastRequestConfig contains configuration for the "least_request" policy type
+type LeastRequestConfig struct {
+	// ChoiceCount determines the number of random healthy hosts from which to select the one with the least requests.
+	ChoiceCount uint32 `json:",omitempty" alias:"choice_count"`
+}
+
+// HashPolicy is a list of hash policies to use for hashing load balancing algorithms.
+// Hash policies are evaluated individually and combined such that identical lists
+// result in the same hash.
+// If no hash policies are present, or none are successfully evaluated,
+// then a random backend host will be selected.
+type HashPolicy struct {
+	// Field is the attribute type to hash on.
+	// Must be one of "header","cookie", or "query_parameter".
+	// Cannot be specified along with SourceIP.
+	Field string `json:",omitempty"`
+
+	// FieldMatchValue is the value to hash.
+	// ie. header name, cookie name, URL query parameter name
+	// Cannot be specified along with SourceIP.
+	FieldMatchValue string `json:",omitempty" alias:"field_value"`
```

Comment:
```suggestion
	FieldValue string `json:",omitempty" alias:"field_value"`
```
For the `alias` thing to work the natural name of the field has to be the camel-case version of the snake-case thing in the alias struct tag.

Reply:
I don't think they have to match, for example these two fields use different names than the canonical one: https://github.com/hashicorp/consul/blob/master/agent/xds/config.go#L98-L101 (with `envoy_` prefix).

Reply:
Edited so that it matches the alias